miniapp/miniapp_band_to_tridiag.cpp

```diff
@@ -0,0 +1,195 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include <iostream>
+#include <limits>
+
+#include <pika/init.hpp>
+#include <pika/program_options.hpp>
+#include <pika/runtime.hpp>
+
+#include "dlaf/common/format_short.h"
+#include "dlaf/common/index2d.h"
+#include "dlaf/common/range2d.h"
+#include "dlaf/common/timer.h"
+#include "dlaf/communication/communicator_grid.h"
+#include "dlaf/communication/init.h"
+#include "dlaf/eigensolver/band_to_tridiag.h"
+#include "dlaf/init.h"
+#include "dlaf/matrix/copy.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/matrix/matrix_mirror.h"
+#include "dlaf/miniapp/dispatch.h"
+#include "dlaf/miniapp/options.h"
+#include "dlaf/types.h"
+
+namespace {
+using dlaf::Backend;
+using dlaf::DefaultDevice_v;
+using dlaf::Device;
+using dlaf::GlobalElementSize;
+using dlaf::Matrix;
+using dlaf::SizeType;
+using dlaf::TileElementSize;
+using dlaf::comm::Communicator;
+using dlaf::comm::CommunicatorGrid;
+using dlaf::common::Ordering;
+using dlaf::matrix::MatrixMirror;
+
+struct Options
+    : dlaf::miniapp::MiniappOptions<dlaf::miniapp::SupportReal::Yes, dlaf::miniapp::SupportComplex::Yes> {
+  SizeType m;
+  SizeType mb;
+  SizeType b;
+  blas::Uplo uplo;
+
+  Options(const pika::program_options::variables_map& vm)
+      : MiniappOptions(vm), m(vm["matrix-size"].as<SizeType>()), mb(vm["block-size"].as<SizeType>()),
+        b(vm["band-size"].as<SizeType>()), uplo(dlaf::miniapp::parseUplo(vm["uplo"].as<std::string>())) {
+    DLAF_ASSERT(m > 0, m);
+    DLAF_ASSERT(mb > 0, mb);
+    DLAF_ASSERT(b > 0 && mb % b == 0, b, mb);
+
+    DLAF_ASSERT(grid_rows * grid_cols == 1,
+                "Error! Distributed is not avilable yet. "
+                "Please rerun with both --grid-rows and --grid-cols set to 1");
+
+    if (backend == Backend::GPU) {
+      std::cerr << "Warning! Backend GPU is not implemented, Using MC backend starting from GPU memory."
+                << std::endl;
+    }
+
+    if (do_check != dlaf::miniapp::CheckIterFreq::None) {
+      std::cerr << "Warning! At the moment result checking it is not implemented." << std::endl;
+      do_check = dlaf::miniapp::CheckIterFreq::None;
+    }
+  }
+
+  Options(Options&&) = default;
+  Options(const Options&) = default;
+  Options& operator=(Options&&) = default;
+  Options& operator=(const Options&) = default;
+};
+}
+
+struct BandToTridiagMiniapp {
+  template <Backend backend, typename T>
+  static void run(const Options& opts) {
+    using MatrixMirrorType = MatrixMirror<T, DefaultDevice_v<backend>, Device::CPU>;
+    using HostMatrixType = Matrix<T, Device::CPU>;
+    using ConstHostMatrixType = Matrix<const T, Device::CPU>;
+
+    if (opts.grid_rows * opts.grid_cols != 1)
+      DLAF_UNIMPLEMENTED("Distributed implementation not available yet.");
```

Comment:
Is this and the other `DLAF_UNIMPLEMENTED` sufficiently covered by the asserts in the `Options` struct? I don't really mind having these here as well since it'll probably help avoid mistakes. Just wanted to check if it's intentional.

Reply:
Yes, the assert in the options is used to give an immediate feedback.
uninmplemented is there to show that the implementation of the `run` member function is broken if the constrain is not satisfied.

Similarly, for wrong sizes we have the assert in option, but we have other checks around in the code.